server-mux.go

```diff
@@ -212,3 +280,83 @@ func longestWord(strings []string) int {
 
 	return maxLen
 }
+
+// Close initiates the graceful shutdown
+func (m *MuxServer) Close() error {
+	m.mu.Lock()
+	if !m.closed {
+		m.Server.SetKeepAlivesEnabled(false)
+		m.closed = true
+		m.mu.Unlock()
+		if err := m.listener.Close(); err != nil {
+			return err
+		}
+
+		// force connections to close after timeout
+		wait := make(chan struct{})
+		go func() {
+			m.mu.Lock()
+			for c, st := range m.conns {
+				// Force close any idle and new connections.
+				if st == http.StateIdle || st == http.StateNew {
+					c.Close()
+				}
+			}
+			m.mu.Unlock()
+
+			// Wait for all connections to be gracefully closed
+			m.wg.Wait()
+			close(wait)
+		}()
+
+		select {
+		case <-time.After(m.GracefulTimeout):
+			return nil
+		case <-wait:
+			return nil
+		}
+	} else {
+		return errors.New("Server has been closed")
+	}
+}
+
+func (m *MuxServer) connState() {
+	oldHook := m.Server.ConnState
+	m.Server.ConnState = func(c net.Conn, cs http.ConnState) {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		switch cs {
```

Comment:
ConnState hook was used to store any previously setup ConnState function and then call it from within the new one. I removed this since it doesn't seem like anyone other than this function will be adjusting the ConnState.